src/net.h

```diff
@@ -866,16 +863,14 @@ class CNode
 
     CNode* AddRef()
     {
-        LOCK(cs_nRefCount);
         nRefCount++;
         return this;
     }
 
     void Release()
     {
-        LOCK(cs_nRefCount);
-        nRefCount--;
-        assert(nRefCount >= 0);
+        int newRefCount = --nRefCount;
+        assert(newRefCount >= 0);
```

Comment:
Not sure I get it... why introducing `newRefCount` here?

Reply:
Because the `assert(nRefCount >= 0)` was checking the result of the decrement and as everything was protected by the CS it was atomic. If I'd simply remove the lock and leave everything else as it is, there would be a race condition introduced because nRefCount would not necessarily contain the result of the decrement if anyone calls another Release or an AddRef in between.

That's also the reason for the pre decrement operator instead of the post increment operator. We want the value of nRefCount after the decrement and not the value before.

Reply:
Why are we removing the lock then? :D

I assumed that it's ok to remove the lock because both `->Release()`s are already protected via `cs_vNodes` but if we assume that this might not be the case we shouldn't remove lock at all imo.

Reply:
... or we should just drop `assert` as well

Reply:
Because the lock was introduced at a time when nRefCount was a simple int and now its an atomic (due to backporting).

I just re-checked when AddRef/Release are called. CConnman only calls AddRef() once when the CNode object is created and Release() once in ThreadSocketHandler when fDisconnect is set to true. The AddRef is not protected by `cs_vNodes` but that's ok because the node object is not part of the nodes vector yet. The Release() call is protected by `cs_vNodes` and because the object is removed from vNodes before it is called, we are protected against Release() being called multiple times.

If it'd be only this, everything would be fine and we probably wouldn't even need the assert() in Release, which would also remove the need for newRefCount...the code would be straight forward. That's how it is in Bitcoin, so they don't have the assert.

But we also have CopyNodeVector and ReleaseNodeVector, and as I understand the code, this is where the assert protects us, as external callers are more likely to make an error (calling ReleaseNodeVector twice).

Hmm but now I'm thinking if it's really that important to assert the result of the decrement (instead of just `assert(nRefCount >= 0);`, as it was before). In case of a race, it would crash nevertheless, but it would be more random which of the 2 racing threads would crash. Not sure if anything is gained from making sure the second thread crashes ;)

Reply:
Oh and I'm actually perfectly fine with removing the assert completely. The amount of CopyNodeVectorCalls gets less and less or should at least not increase anymore as we now have ForNode.

Reply:
Yep. I'd just re-align this code with bitcoin again ~and let it crash~ :D

Reply:
If this crashes (again) i will be sending you coredumps again :D

Reply:
Force pushed with assert being completely removed